### PR TITLE
Submodules init and update

### PR DESCRIPTION
**Pull request by mcuadros**
This is the implementation of `git submodule init` and `git submodule update`, this implementation has required make some changes non direct related:

- storage: git.Storer move to storage.Storer and module handling 	498dbf7, due to the requirement of make reference to the `storage.Storer` as part of itself was moved to another package in other to avoid a cyclic dependency
- plumbing/object: allow TreeIter return SubModule entries b18d649 the submodule were ignored, now are being returned without problem
- format/index: sort the Entries before encode 940a16c, this was a fix since a Index entries should be save in order by Name
- config: adding Config.Core.Worktree f8b5557
- Repository.Init now handles non-standard .git location` 65351f8, it creates a `.git` file in the root of the worktree and sets Config.Worktree when the .git folder is not in a standard location

Once this is merge, I will do another PR renaming `config.RemoteConfig` to `config.Remote` to mimic `config.Submodule`

### CODE REVIEW

File: _examples/clone/main.go [hunk 1]
```diff
@@ -14,11 +14,11 @@ func main() {
...
 
 	// Clone the given repository to the given directory
-	Info("git clone %s %s", url, directory)
+	Info("git clone %s %s --recursive", url, directory)
 
 	r, err := git.PlainClone(directory, false, &git.CloneOptions{
-		URL:   url,
-		Depth: 1,
+		URL:               url,
+		RecurseSubmodules: git.DefaultRecursivity,
```
Comment (smola) [inline _examples/clone/main.go]: I would change the name from `DefaultRecursivity` to `DefaultRecurseModules` or `DefaultSubmoduleRecursivity` or something like that, so that it does not get confused with other parameters related with recursivite of non-submodule stuff (e.g. reference resolution).

File: options.go [hunk 2]
```diff
@@ -148,3 +164,16 @@ func (o *PushOptions) Validate() error {
 
 	return nil
 }
+
+// SubmoduleUpdateOptions describes how a submodule update should be performed
+type SubmoduleUpdateOptions struct {
+	// Init initializes the submodules recorded in the index
+	Init bool
+	// NoFetch tell to the update command to don’t fetch new objects from the
```
Comment (smola) [inline options.go]: `to don’t` -> `to not`

File: plumbing/format/index/encoder.go [hunk 3]
```diff
@@ -139,3 +142,9 @@ func (e *Encoder) padEntry(wrote int) error {
 func (e *Encoder) encodeFooter() error {
 	return binary.Write(e.w, e.hash.Sum(nil))
 }
+
+type ByName []Entry
```
Comment (smola) [inline plumbing/format/index/encoder.go]: private type?

File: storage/test/storage_suite.go [hunk 4]
```diff
@@ -21,6 +22,8 @@ type Storer interface {
 	storer.ShallowStorer
 	storer.IndexStorer
 	config.ConfigStorer
+
```
Comment (alcortesm) [inline storage/test/storage_suite.go]: remove extra line here.

File: storage/test/storage_suite.go [hunk 5]
```diff
@@ -326,6 +329,16 @@ func (s *BaseStorageSuite) TestSetConfigInvalid(c *C) {
 	c.Assert(err, NotNil)
 }
 
+func (s *BaseStorageSuite) TestModule(c *C) {
+	storer, err := s.Storer.Module("foo")
+	c.Assert(err, IsNil)
+	c.Assert(storer, NotNil)
+
+	storer, err = s.Storer.Module("foo")
```
Comment (alcortesm) [inline storage/test/storage_suite.go]: What is the point in calling Module twice?
(explain this in the name of the test)

Actually what is the point of this whole test?

File: storage/memory/storage.go [hunk 6]
```diff
@@ -227,3 +230,13 @@ func (s *ShallowStorage) SetShallow(commits []plumbing.Hash) error {
...
 }
+
+type ModuleStorage map[string]*Storage
+
+func (s ModuleStorage) Module(name string) (storage.Storer, error) {
+	if _, ok := s[name]; !ok {
+		s[name] = NewStorage()
+	}
+
+	return s[name], nil
```
Comment (alcortesm) [inline storage/memory/storage.go]: Don't look up twice in the map, use a variable to store the results of the first lookup.

File: storage/storer.go [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+type Storer interface {
+	storer.EncodedObjectStorer
+	storer.ReferenceStorer
+	storer.ShallowStorer
+	storer.IndexStorer
+	config.ConfigStorer
+	ModuleStorer
+}
+
+type ModuleStorer interface {
```
Comment (alcortesm) [inline storage/storer.go]: This needs to be explained, What if the module doesn't exists?

File: repository.go [hunk 8]
```diff
@@ -270,11 +265,13 @@ func (r *Repository) clone(o *CloneOptions) error {
 		Progress: o.Progress,
 	})
 	if err != nil {
+
```
Comment (alcortesm) [inline repository.go]: remove empty line.

File: repository.go [hunk 9]
```diff
@@ -270,11 +265,13 @@ func (r *Repository) clone(o *CloneOptions) error {
...
 		return err
 	}
 
 	head, err := storer.ResolveReference(remoteRefs, o.ReferenceName)
 	if err != nil {
+
```
Comment (alcortesm) [inline repository.go]: remove empty line.

File: repository.go [hunk 10]
```diff
@@ -283,12 +280,33 @@ func (r *Repository) clone(o *CloneOptions) error {
 	}
 
 	if err := r.updateWorktree(); err != nil {
+		fmt.Println("q", err)
```
Comment (alcortesm) [inline repository.go]: leftover?

File: repository.go [hunk 11]
```diff
@@ -283,12 +280,33 @@ func (r *Repository) clone(o *CloneOptions) error {
...
+	if err != nil {
+		return err
+	}
+
+	s, err := w.Submodules()
+	if err != nil {
+		return err
+	}
+
+	return s.Init()
```
Comment (alcortesm) [inline repository.go]: Should this have a recursive flag, so the init can be done recursively?

File: worktree.go [hunk 12]
```diff
@@ -167,6 +170,61 @@ func (w *Worktree) getMode(fi billy.FileInfo) os.FileMode {
 	return object.FileMode
 }
 
+const gitmodulesFile = ".gitmodules"
+
+func (w *Worktree) Submodules() (Submodules, error) {
+	l := make(Submodules, 0)
```
Comment (alcortesm) [inline worktree.go]: make this after calling readGitmodulesFile, and make it with the size of the m.Submodules,

File: worktree.go [hunk 13]
```diff
@@ -167,6 +170,61 @@ func (w *Worktree) getMode(fi billy.FileInfo) os.FileMode {
...
+	}
+
+	input, err := ioutil.ReadAll(f)
+	if err != nil {
+		return nil, err
+	}
+
+	m := config.NewModules()
+	return m, m.Unmarshal(input)
+
```
Comment (alcortesm) [inline worktree.go]: Remove empty line.

File: worktree.go [hunk 14]
```diff
@@ -167,6 +170,61 @@ func (w *Worktree) getMode(fi billy.FileInfo) os.FileMode {
 	return object.FileMode
 }
 
+const gitmodulesFile = ".gitmodules"
+
+func (w *Worktree) Submodules() (Submodules, error) {
```
Comment (alcortesm) [inline worktree.go]: document it, especially why it can return (empty slice and nil) and what does it means. Is it an error if the .gitmodules is empty?

File: plumbing/object/file.go [hunk 15]
```diff
@@ -81,7 +81,7 @@ func (iter *FileIter) Next() (*File, error) {
 			return nil, err
 		}
 
-		if entry.Mode.IsDir() {
+		if entry.Mode.IsDir() || entry.Mode == SubmoduleMode {
```
Comment (alcortesm) [inline plumbing/object/file.go]: Invert the logic here. Just continue if it is not a file, otherwise we are always adding exceptions as they come along.

File: plumbing/object/file_test.go [hunk 16]
```diff
@@ -247,3 +247,29 @@ func (s *FileSuite) TestFileIter(c *C) {
...
+	tree, err := commit.Tree()
+	c.Assert(err, IsNil)
+
+	expected := []string{
+		".gitmodules",
+	}
+
+	var count int
+	i := tree.Files()
+	i.ForEach(func(f *File) error {
```
Comment (alcortesm) [inline plumbing/object/file_test.go]: There are several things that are wrong with this test:

- it is too complex to test that the obtained must be a single file.
- If Files return an empty list of file, I'm pretty sure the test will pass, so we will never know if submodules are ignored.
- the name of the test does not explain what it is doing (check that submodules are ignored)
- the test is not calling the function that has the logic under test (FileIter.Next)

File: plumbing/object/file_test.go [hunk 17]
```diff
@@ -247,3 +247,29 @@ func (s *FileSuite) TestFileIter(c *C) {
 
 	c.Assert(count, Equals, 1)
 }
+
+func (s *FileSuite) TestFileIterSubmodule(c *C) {
+	st, err := filesystem.NewStorage(fixtures.ByTag("submodule").One().DotGit())
```
Comment (alcortesm) [inline plumbing/object/file_test.go]: please, comment what repository are you using for this test, it is very hard to know what should be inside the commit below otherwise.

File: plumbing/object/tree_test.go [hunk 18]
```diff
@@ -262,6 +265,41 @@ func (s *TreeSuite) TestTreeWalkerNextNonRecursive(c *C) {
 	c.Assert(count, Equals, 8)
 }
 
+func (s *TreeSuite) TestTreeWalkerNextSubmodule(c *C) {
+	st, err := filesystem.NewStorage(fixtures.ByTag("submodule").One().DotGit())
```
Comment (alcortesm) [inline plumbing/object/tree_test.go]: please, comment what repository are you using for this test, it is very hard to know what should be inside the commit below otherwise.

File: plumbing/object/tree_test.go [hunk 19]
```diff
@@ -262,6 +265,41 @@ func (s *TreeSuite) TestTreeWalkerNextNonRecursive(c *C) {
...
+	c.Assert(err, IsNil)
+
+	expected := []string{
+		".gitmodules",
+		"basic",
+		"itself",
+	}
+
+	var count int
+	walker := NewTreeWalker(tree, true)
```
Comment (alcortesm) [inline plumbing/object/tree_test.go]: close the walker, or make NewTreeWalker return the close closure so we never forget about it again.

File: plumbing/format/index/encoder.go [hunk 20]
```diff
@@ -61,6 +62,8 @@ func (e *Encoder) encodeHeader(idx *Index) error {
 }
 
 func (e *Encoder) encodeEntries(idx *Index) error {
+	sort.Sort(ByName(idx.Entries))
```
Comment (alcortesm) [inline plumbing/format/index/encoder.go]: Why? please explain and add tests if this is a bug.

File: storage/filesystem/internal/dotgit/dotgit.go [hunk 21]
```diff
@@ -390,6 +390,11 @@ func (d *DotGit) readReferenceFile(refsPath, refFile string) (ref *plumbing.Refe
 	return plumbing.NewReferenceFromStrings(refFile, line), nil
 }
 
+// Module return a billy.Filesystem poiting to the module folder
+func (d *DotGit) Module(name string) billy.Filesystem {
```
Comment (alcortesm) [inline storage/filesystem/internal/dotgit/dotgit.go]: Add tests.

File: storage/filesystem/module.go [hunk 22]
```diff
@@ -0,0 +1,14 @@
...
+import (
+	"srcd.works/go-git.v4/storage"
+	"srcd.works/go-git.v4/storage/filesystem/internal/dotgit"
+)
+
+type ModuleStorage struct {
+	dir *dotgit.DotGit
+}
+
+func (s *ModuleStorage) Module(name string) (storage.Storer, error) {
```
Comment (alcortesm) [inline storage/filesystem/module.go]: Add tests and comments.

File: storage/filesystem/module.go [hunk 23]
```diff
@@ -0,0 +1,14 @@
+package filesystem
+
+import (
+	"srcd.works/go-git.v4/storage"
+	"srcd.works/go-git.v4/storage/filesystem/internal/dotgit"
+)
+
+type ModuleStorage struct {
```
Comment (alcortesm) [inline storage/filesystem/module.go]: Do we really need this type?, can we just make dotgit implement `storage.ModuleStorer` instead.

File: config/config.go [hunk 24]
```diff
@@ -37,6 +37,8 @@ type Config struct {
 		// IsBare if true this repository is assumed to be bare and has no
 		// working directory associated with it
 		IsBare bool
+		// Worktree is the path to the root of the working tree
+		Worktree string
```
Comment (alcortesm) [inline config/config.go]: change name to `WorkTtree`.

Add full stop at the end of the phrase.

We should call it `WorkTreePath`.

File: config/config.go [hunk 25]
```diff
@@ -76,6 +78,7 @@ const (
 	fetchKey      = "fetch"
 	urlKey        = "url"
 	bareKey       = "bare"
+	worktreeKey   = "worktree"
```
Comment (alcortesm) [inline config/config.go]: Change this to `workTreeKey`.

File: repository.go [hunk 26]
```diff
@@ -57,9 +58,75 @@ func Init(s storage.Storer, worktree billy.Filesystem) (*Repository, error) {
...
 	if worktree == nil {
 		r.setIsBare(true)
+		return r, nil
 	}
 
-	return r, nil
+	return r, setWorktreeAndStoragePaths(r, worktree)
+}
+
+func setWorktreeAndStoragePaths(r *Repository, worktree billy.Filesystem) error {
```
Comment (alcortesm) [inline repository.go]: Capitalize `tree` in the function name.

File: repository.go [hunk 27]
```diff
@@ -57,9 +58,75 @@ func Init(s storage.Storer, worktree billy.Filesystem) (*Repository, error) {
...
+	if err != nil {
+		return err
+	}
+
+	defer f.Close()
+	_, err = fmt.Fprintf(f, "gitdir: %s\n", path)
+	return err
+}
+
+func setConfigWorktree(r *Repository, worktree, storage billy.Filesystem) error {
```
Comment (alcortesm) [inline repository.go]: Capitalize `tree` in function name.

File: submodule.go [hunk 28]
```diff
@@ -1,33 +1,79 @@
...
 )
 
+// Submodule a submodule allows you to keep another Git repository in a
+// subdirectory of your repository.
 type Submodule struct {
-	Name   string
-	Branch string
-	URL    string
-
+	m *config.Submodule
```
Comment (alcortesm) [inline submodule.go]: Why is it called m?

File: submodule.go [hunk 29]
```diff
@@ -1,33 +1,79 @@
...
+	w *Worktree
```
Comment (alcortesm) [inline submodule.go]: Change this to WorkTree.

File: submodule.go [hunk 30]
```diff
@@ -1,33 +1,79 @@
...
+	// r is the submodule repository
 	r *Repository
 }
 
+// Config returns the submodule config
+func (s *Submodule) Config() *config.Submodule {
```
Comment (alcortesm) [inline submodule.go]: Why not make m public instead of a getter?

File: worktree.go [hunk 31]
```diff
@@ -210,6 +210,22 @@ func (w *Worktree) getMode(fi billy.FileInfo) os.FileMode {
 
 const gitmodulesFile = ".gitmodules"
 
+// Submodule returns the submodule with the given name
```
Comment (alcortesm) [inline worktree.go]: Explains what happens if it does not exists.

File: worktree.go [hunk 32]
```diff
@@ -210,6 +210,22 @@ func (w *Worktree) getMode(fi billy.FileInfo) os.FileMode {
...
+
+	c, ok := m.Submodules[name]
+	if !ok {
+		return nil, ErrSubmoduleNotFound
+	}
+
+	return w.newSubmodule(c)
+}
+
+// Submodules returns all the available submodules
```
Comment (alcortesm) [inline worktree.go]: Explain that this also init the submodules and why?

File: config/config.go [hunk 33]
```diff
@@ -40,8 +40,10 @@ type Config struct {
 		// Worktree is the path to the root of the working tree
 		Worktree string
 	}
-	// Remote list of repository remotes
+	// Remotes list of repository remotes
```
Comment (alcortesm) [inline config/config.go]: End the sentence with a full stop and add a verb.

File: config/config.go [hunk 34]
```diff
@@ -40,8 +40,10 @@ type Config struct {
...
 	Remotes map[string]*RemoteConfig
+	// Submodules list of repository submodules
```
Comment (alcortesm) [inline config/config.go]: End the sentence with a full stop and add a verb.

Comment (alcortesm) [inline config/config.go]: Explain what is the key to the map.

File: config/config.go [hunk 35]
```diff
@@ -118,10 +123,21 @@ func (c *Config) unmarshalRemotes() error {
 	return nil
 }
 
+func (c *Config) unmarshalSubmodules() {
+	s := c.raw.Section(submoduleSection)
+	for _, sub := range s.Subsections {
+		m := &Submodule{}
+		m.unmarshal(sub)
+
```
Comment (alcortesm) [inline config/config.go]: remove empty line.

File: options.go [hunk 36]
```diff
@@ -9,9 +9,18 @@ import (
 	"srcd.works/go-git.v4/plumbing/transport"
 )
 
+// SubmoduleRescursivity defines how depth will affect any submodule recursive
+// operation
+type SubmoduleRescursivity int
```
Comment (alcortesm) [inline options.go]: alternative name: `SubmoduleRecursionDepth`, explain what negative values will do and how to make it infinite.

File: options.go [hunk 37]
```diff
@@ -9,9 +9,18 @@ import (
...
+
 const (
 	// DefaultRemoteName name of the default Remote, just like git command
 	DefaultRemoteName = "origin"
+
+	// NoRecursivity disables the recursion for a submodule operation
+	NoRecursivity SubmoduleRescursivity = 0
```
Comment (alcortesm) [inline options.go]: alternative: `NonRecursive`.

File: options.go [hunk 38]
```diff
@@ -9,9 +9,18 @@ import (
...
+	// DefaultRecursivity allow recursion in a submodule operation
+	DefaultRecursivity SubmoduleRescursivity = 10
```
Comment (alcortesm) [inline options.go]: alternative: `DefaultSubmoduleRecursionDepth`.

File: options.go [hunk 39]
```diff
@@ -71,6 +80,9 @@ type PullOptions struct {
 	Depth int
 	// Auth credentials, if required, to use with the remote repository
 	Auth transport.AuthMethod
+	// RecurseSubmodules controls if new commits of all populated submodules
+	// should be fetched too
```
Comment (alcortesm) [inline options.go]: End the sentence with a full stop.

File: options.go [hunk 40]
```diff
@@ -152,3 +164,16 @@ func (o *PushOptions) Validate() error {
 
 	return nil
 }
+
+// SubmoduleUpdateOptions describes how a submodule update should be performed
```
Comment (alcortesm) [inline options.go]: Add full stop.

File: options.go [hunk 41]
```diff
@@ -152,3 +164,16 @@ func (o *PushOptions) Validate() error {
...
+type SubmoduleUpdateOptions struct {
+	// Init initializes the submodules recorded in the index
```
Comment (alcortesm) [inline options.go]: Add full stop.

Comment (alcortesm) [inline options.go]: This needs much better explanation. It is a bool, but does not make sense with the comment.

File: storage/test/storage_suite.go (hunk 5 above)
Comment (mcuadros) [inline storage/test/storage_suite.go]: test that returns a module, that exists always, calling it twice looks an error

File: doc.go [hunk 42]
```diff
@@ -7,4 +7,4 @@
 // It is highly extensible, we have been following the open/close principle in
 // its design to facilitate extensions, mainly focusing the efforts on the
 // persistence of the objects.
-package git
+package git // import "srcd.works/go-git.v4"
```
Comment (mcuadros) [inline doc.go]: yup

File: plumbing/object/file.go (hunk 15 above)
Comment (mcuadros) [inline plumbing/object/file.go]: Can you elaborate?

File: plumbing/object/file_test.go (hunk 17 above)
Comment (mcuadros) [inline plumbing/object/file_test.go]: true, if an object is referenced, we should use the url and not a tag

File: plumbing/object/file_test.go (hunk 16 above)
Comment (mcuadros) [inline plumbing/object/file_test.go]: > it is too complex to test that the obtained must be a single file.
It's asserting the full content of the commit, in this case is only one file that is not a submodule

> If Files return an empty list of file, I'm pretty sure the test will pass, so we will never know if submodules are ignored.
Nope, we have that is returned only one file, not less or more.

> the name of the test does not explain what it is doing (check that submodules are ignored)
I can change the name, but usually for me is explaining that is walking in a commit with submodule, I am not want to explain the logic behind.

> the test is not calling the function that has the logic under test (FileIter.Next)
ForEach calls Next

File: plumbing/format/index/encoder.go (hunk 20 above)
Comment (mcuadros) [inline plumbing/format/index/encoder.go]: this is already covered by the tests, I introduced another entry in the tests

File: storage/filesystem/internal/dotgit/dotgit.go (hunk 21 above)
Comment (mcuadros) [inline storage/filesystem/internal/dotgit/dotgit.go]: done

File: storage/filesystem/module.go (hunk 22 above)
Comment (mcuadros) [inline storage/filesystem/module.go]: Is already covered by the tests. And the documentation is on the interface as in the other methods

File: storage/filesystem/module.go (hunk 23 above)
Comment (mcuadros) [inline storage/filesystem/module.go]: This is the pattern we are using. I prefer keep dotgit isolate

File: config/config.go (hunk 24 above)
Comment (mcuadros) [inline config/config.go]: worktree is a single word, is a concept. I think that is enough to explained that is a path, since a worktree and is a string.

File: config/config.go (hunk 35 above)
Comment (mcuadros) [inline config/config.go]: why?

File: options.go (hunk 36 above)
Comment (mcuadros) [inline options.go]: I just change the time to uint, infinity by itself is not allowed, a number should be given always, even if is very high

File: storage/test/storage_suite.go (hunk 5 above)
Comment (alcortesm) [inline storage/test/storage_suite.go]: I don't understand your answer.

Can you explain what you are testing in the name of the test or on a comment?

File: plumbing/object/file.go (hunk 15 above)
Comment (alcortesm) [inline plumbing/object/file.go]: Now, this is how it works:

The `FileIter.Next` method returns `Files` by going over all the elements in the tree, ignoring Dirs, and now (added in this PR) Submodules.

This is what I propose, instead:

The `FileIter.Next` method returns `Files` by going over all the elements in the tree, ignoring everything that is not a file (without naming explicitly what are not files).

This way we don't have to change the code everytime we support a new TreeEntry mode.

So instead of:
```
if entry.Mode.IsDir() || entry.Mode == SubmoduleMode {
    continue
}
```

I suggest:

```
if !entry.Mode.IsFile() && !entry.Mode.IsOldFileGroupWriteable() {
    continue
}
```

File: config/config.go (hunk 35 above)
Comment (alcortesm) [inline config/config.go]: I don't know why, I don't understand the coding standards we use in the project, I'm just trying to make your code look like the rest of the code in the project. We usually only insert empty lines before a return or to separate "code paragraphs".
